WIP: Involve serial console for preventing failing needle

clanig opened the pull request:
This PR shifts things to be done on the `rescue.target` to the serial console to improve the consistency of the results.
The other code is left in there as fallback.

- Related ticket: https://progress.opensuse.org/issues/36280
- Verification run:
![after](https://user-images.githubusercontent.com/31621613/41475320-c594f744-70be-11e8-927f-bef9928a2f81.jpg)

rwx788 commented on lib/btrfs_test.pm:
```diff
@@ -40,13 +41,51 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+    # serial console prints text style as ascii chars in PS1 by default.
+    my $escseq = '[\]\[\(\)\e\d\w\s]*';
+
     if (script_run('! systemctl is-active dbus')) {
```
What about using early return here to reduce nested if clauses:
```
return unless script_run('! systemctl is-active dbus');
```

rwx788 commented on lib/btrfs_test.pm:
```diff
@@ -40,13 +41,51 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+            wait_serial 'sed -i \'s\@\^StandardInput=\.\*\$\@StandardInput=tty\\\\nTTYPath=\'\\$\(tty\)\'\@g\' \/usr\/lib\/systemd\/system\/rescue\.service';
+
+            type_string "\n";
+            wait_serial("# ");
```
I'm not really sure, but can we use script_run for execution here instead of typing and waiting for serial? Or maybe create some wrapper which does these two calls and wait for the same string which was typed by default.

clanig commented:
@rwx788 Thank you for the suggestions. I have:
- [x] Replaced the if-block with the return statement as suggested.
- [x] Replaced the parts with script_run that were replaceable as suggested.
- [x] Outsourced the path to the service file to a variable.

When there is something more to add, I am always happy for feedback.

okurz commented on lib/btrfs_test.pm:
```diff
@@ -40,7 +41,34 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+    if (serial_terminal::select_virtio_console()) {
+
+        # Replace StandardInput in rescue.service file and add TTYPath.
+        script_run "cp $rescue_service_config_path $rescue_service_config_path.sav";
```
I think a more common way is to provide a [replacement unit file](https://wiki.archlinux.org/index.php/systemd#Replacement_unit_files) by creating the file /etc/systemd/system/rescue.service or create a drop-in file. Can you use `assert_script_run` by default? We don't want to continue the test when even this copy command fails, right? By default it is suggested to use `assert_script_run` unless you want to explicitly check the return value, e.g. as you do in line 49

clanig replied:
Thank you for your hint! I wasn't aware of this option. I will try to implement it the mentioned way.

okurz commented on lib/btrfs_test.pm:
```diff
@@ -40,7 +41,34 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+        type_string 'PS1="# "; export TERM=dumb; stty cols 2048';
+        wait_serial 'PS1="# "; export TERM=dumb; stty cols 2048';
+        type_string "\n";
+        wait_serial "# ";
```
Same as @rwx788 I don't understand why we would need to repeat each "type_string" and "wait_serial" call twice each time. Can you not use simply `assert_script_run`?

clanig replied:
In this case I had indeed issues that this wasn't working reliably. But I will have another try what I can do to get rid of the overhead.

okurz commented on lib/btrfs_test.pm:
```diff
@@ -57,7 +85,33 @@ move from rescue to default target, logs us out. Die if DBus is active at this p
+    type_string "tty";
+    wait_serial "tty";
+    type_string "\n";
+    if (wait_serial('\/dev\/hvc\d+', 5)) {
```
Is there a specific reason why this must appear within 5 seconds? Better use the default timeout unless you want to catch something that is really time critical.

clanig replied:
There is no reason for it to having to appear within 5 seconds. I did that because it appeared a huge delay for simple serial console I/O.
But I will remove it because it's not necessary and might also look confusing.

clanig commented:
@okurz Thank you for reviewing my PR and giving me useful feedback.

Indeed I have no info whether this issue can still happen in recent tests. But improving the accuracy of the test makes sense to me when we have the indication that it might be irritated.

clanig commented:
I have updated this PR.
* `assert_script_run` is used whenever possibe.
* The way how the service file is changed has been altered.
* Unfortunately it is not possible to run `type_string 'PS1="# "; export TERM=dumb; stty cols 2048';` differently because it looks like script_run expects a hashtag before execution and this is achieved this command.
* I have also removed the old way how it was done before.

clanig commented:
@okurz Thanks for your reply. Your concerns are valid, especially regarding the issues with systems that don't provide a virtio serial connection.
I will leave this a bit open to see whether I can find a reasonable solution. However, it seems to be difficult to find one for this case atm.

richiejp commented on lib/btrfs_test.pm:
```diff
@@ -40,14 +41,35 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+
+    return unless script_run('! systemctl is-active dbus');
+    if (!serial_terminal::select_virtio_console()) {
+        die 'Selection of virtio console failed.';
```
I don't think it is necessary to die here. You should be able to adapt the test for both vnc and serial terminals.

richiejp replied:
to be clear; you can probably just use the existing code if select_virtio_console returns zero. Or you can branch later with is_serial_terminal.

richiejp commented on lib/btrfs_test.pm:
```diff
@@ -40,14 +41,35 @@ way to get DBus-less environment is to enter rescue.target via systemctl.
+    wait_serial 'PS1="# "; export TERM=dumb; stty cols 2048';
+    type_string "\n";
+    wait_serial "# ";
+    $self->set_standard_prompt('root');
```
This part is repeating some code in lib/serial_terminal.pm. You can refactor that file to add a function which sets the prompt and use it here and in the normal login function for virtio-terminal. You just need to verify that the LTP tests still run after the refactoring.

richiejp commented on lib/btrfs_test.pm:
```diff
@@ -57,16 +79,24 @@ move from rescue to default target, logs us out. Die if DBus is active at this p
+    wait_serial "tty";
+    type_string "\n";
+    if (!wait_serial '\/dev\/hvc\d+') {
+        die 'This subroutine needs a virtio console being selected.';
```
use `is_serial_terminal`

fyi hvc does not necessarily use virtio.

richiejp commented on lib/btrfs_test.pm:
```diff
@@ -57,16 +79,24 @@ move from rescue to default target, logs us out. Die if DBus is active at this p
+    wait_serial "# ";
+    type_string "systemctl default";
+    wait_serial "systemctl default";
+    type_string "\n";
```
you can probably use `assert_script_run` here.

richiejp commented:
I think this is a good idea in principal as matching needles to text is less reliable and slow. You mainly just need to remove some duplicate code and keep the non-serial-terminal code path.

clanig commented:
@richiejp 
Thank you for the detailed feedback. I will implement it and then update this PR again.

clanig commented:
I have installed a new Openqa development environment and I am optimistic to be able to update the PR until the end of next week.
Sorry for the delay.
